{% load static %} {% load static %} {% load crispy_forms_tags %}
{% load rtags %}

<!DOCTYPE html>
<!--[if lt IE 7 ]><html class="ie ie6" lang="en"> <![endif]-->
<!--[if IE 7 ]><html class="ie ie7" lang="en"> <![endif]-->
<!--[if IE 8 ]><html class="ie ie8" lang="en"> <![endif]-->
<!--[if (gte IE 9)|!(IE)]><!-->
<html lang="en">
<!--<![endif]-->

<head>

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <meta name="description" content="">
    <meta name="author" content="">
    <meta name="keywords" content="">


    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <link href="https://gitcdn.github.io/bootstrap-toggle/2.2.2/css/bootstrap-toggle.min.css" rel="stylesheet">
     <script src="https://gitcdn.github.io/bootstrap-toggle/2.2.2/js/bootstrap-toggle.min.js"></script>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">
    <!-- Latest compiled and minified CSS -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
    <!-- Optional theme -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css" integrity="sha384-rHyoN1iRsVXV4nD0JutlnGaslCJuC7uwjduW9SVrLvRYooPp2bWYgmgJQIXwl/Sp" crossorigin="anonymous">
    <!-- Latest compiled and minified JavaScript -->
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script type="text/javascript" async="" src="https://www.google-analytics.com/analytics.js"></script>
    <script async="" src="https://www.googletagmanager.com/gtag/js?id=UA-117917519-1"></script>
    <script>
        window.dataLayer = window.dataLayer || [];
          function gtag(){dataLayer.push(arguments);}
          gtag('js', new Date());

          gtag('config', 'UA-117917519-1');
    </script>
    <meta charset="UTF-8">
    <title>DiNiro</title>
    <link rel="stylesheet" href="{% rstatic 'test_interface/rc-slider@6.1.2.css' %}">
    <link rel="stylesheet" href="{% rstatic 'test_interface/react-select@1.0.0-rc.3.min.css' %}">
    <link rel="stylesheet" href="{% rstatic 'test_interface/react-virtualized@9.9.0.css' %}">
    <link rel="stylesheet" href="{% rstatic 'test_interface/react-virtualized-select@3.1.0.css' %}">
    <link rel="stylesheet" href="{% rstatic 'test_interface/react-dates@12.3.0.css' %}">
    <link rel="stylesheet" href="{% rstatic 'test_interface/STREAM.css' %}">
    <link rel="stylesheet" href="{% rstatic 'test_interface/Loading-State.css' %}">
    <script src="{% rstatic 'test_interface/jquery-3.3.1.min.js' %}"></script>


    <!-- Custom styles for this template -->
    <!link href="jumbotron.css" rel="stylesheet">


    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
      <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->

    <!link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.1/css/all.min.css">
    <link rel="stylesheet" type="text/css" href="{% rstatic 'test_interface/css/all.css' %}">






    <title>Item Name | Documentation by Author Name</title>

    <link rel="shortcut icon" href="{% rstatic 'test_interface/docum/images/favicon.ico' %}" type="image/x-icon">

    <link rel="stylesheet" type="text/css" href="{% rstatic 'test_interface/docum/fonts/font-awesome-4.3.0/css/font-awesome.min.css' %}">
    <link rel="stylesheet" type="text/css" href="{% rstatic 'test_interface/docum/css/stroke.css' %}">
    <link rel="stylesheet" type="text/css" href="{% rstatic 'test_interface/docum/css/bootstrap.css' %}">
    <link rel="stylesheet" type="text/css" href="{% rstatic 'test_interface/docum/css/animate.css' %}">
    <link rel="stylesheet" type="text/css" href="{% rstatic 'test_interface/docum/css/prettyPhoto.css' %}">
    <link rel="stylesheet" type="text/css" href="{% rstatic 'test_interface/docum/css/style.css' %}">

    <link rel="stylesheet" type="text/css" href="{% rstatic 'test_interface/docum/js/syntax-highlighter/styles/shCore.css' %}" media="all">
    <link rel="stylesheet" type="text/css" href="{% rstatic 'test_interface/docum/js/syntax-highlighter/styles/shThemeRDark.css' %}" media="all">

    <!-- CUSTOM -->
    <link rel="stylesheet" type="text/css" href="{% rstatic 'test_interface/docum/css/custom.css' %}">

    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
      <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->
        <style>
        .p {
            color: black;
        }

        .section {
            padding: 0px !important;
            margin-bottom: 50px;
        }
        .switch {
          position: relative;
          display: inline-block;
          width: 60px;
          height: 34px;
        }
        .card1 {
          box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
          transition: 0s;
          width: 100%;
        }
        .switch input {
          opacity: 0;
          width: 0;
          height: 0;
        }
        .slider {
          position: absolute;
          cursor: pointer;
          top: 0;
          left: 0;
          right: 0;
          bottom: 0;
          background-color: deepskyblue;
          -webkit-transition: .4s;
          transition: .4s;
        }

        .slider:before {
          position: absolute;
          content: "";
          height: 26px;
          width: 26px;
          left: 4px;
          bottom: 4px;
          background-color: white;
          -webkit-transition: .4s;
          transition: .4s;
        }

        input:checked + .slider {
          background-color: #2196F3;
        }

        input:focus + .slider {
          box-shadow: 0 0 1px #2196F3;
        }

        input:checked + .slider:before {
          -webkit-transform: translateX(26px);
          -ms-transform: translateX(26px);
          transform: translateX(26px);
        }

        /* Rounded sliders */
        .slider.round {
          border-radius: 34px;
        }

        .slider.round:before {
          border-radius: 50%;
        }
        .center {
          display: block;
          margin-left: auto;
          margin-right: auto;
        }

      .active1 {
          color: #7d7d7d;
          float: left;
          padding: 0px;
          {#color: #9d9d9d;#}
          text-decoration: none !important;
          font-size: 25px;
          position: relative;
          right: 0mm;
          border: 0px solid #73AD21;

        }
      .active2 {
          float: right;
          padding: 0px;
          color: white;
          margin-top: 6px;
          text-decoration: none !important;
          font-size: 18px;
          position: relative;
          left: 0mm;
          margin-right: 0mm;
          border: 0px solid #73AD21;
        }
      .active22, .btn22:hover {
          color: #FFFFFF;
          text-decoration: none !important;
        }
      .selection-with {
         width:300px;
      }
      .graph-container {
            width:100% !important;
            height:100% !important;
      }
      .navbar-custom {
          height: 50px;
      }
      .footer-custom {
          height: 40px;
          padding-top: 10px;
      }
      .stepwizard-step p {
    margin-top: 0px;
    color:#666;
    }
    .stepwizard-row {
        display: table-row;
    }
    .stepwizard {
        display: table;
        width: 100%;
        position: relative;
    }
    .stepwizard-step button[disabled] {
        /*opacity: 1 !important;
        filter: alpha(opacity=100) !important;*/
    }
    .stepwizard .btn.disabled, .stepwizard .btn[disabled], .stepwizard fieldset[disabled] .btn {
        opacity:1 !important;
        color:#bbb;
    }
    .ali{
        height: 100%;!important;
        width: 100%;!important;
        padding-bottom: 0px;!important;
        margin-bottom: 0px;!important;
        min-height: 100%;!important;
    }
    .nomargin{
        height: 100%;!important;
        width: 100%;!important;
        margin-left: 0px;!important
        padding-bottom: 0px;!important;
        margin-top: 0px;!important;
        margin-right: 0px;!important;
        bottom: 0px;!important;
        left: 0px;!important;
        right: 0px;!important;
        top: 0px;!important;
    }
    .stepwizard-row:before {
        top: 28px;
        bottom: 0;
        position: absolute;
        content:" ";
        width: 100%;
        height: 1px;
        background-color: #ccc;
        z-index: 0;
    }
    .stepwizard-step {
        display: table-cell;
        text-align: center;
        position: relative;
    }
    .btn-circle {
        width: 60px;
        height: 60px;
        text-align: center;
        padding: 12px 0;
        font-size: 24px;
        line-height: 1.428571429;
        border-radius: 30px;
    }
    .act {
        text-align: initial;
        border-style: solid;
        border-width: 1mm;
        border-color: deepskyblue;
        min-width: 50%;
        background-color: deepskyblue;
        border-radius: 2px;
        margin-right: 0px;
        color: white
    }
    .pas {
        min-width: 50%;
        background-color: white;
        border-radius: 2px;
        color: #33C3F0;
        border-style: solid;
        border-width: 1mm;
        border-color: deepskyblue;
    }
    </style>


</head>

<body>
    <button onclick="topFunction()" id="myBtn" title="Go to top"><i class="fa fa-chevron-up" aria-hidden="true"></i></button>

    <script>
        var mybutton = document.getElementById("myBtn");
        window.onscroll = function() {scrollFunction()};
        function scrollFunction() {
            if (document.body.scrollTop > 1000 || document.documentElement.scrollTop > 1000) {
                mybutton.style.display = "block";
            } else {
                mybutton.style.display = "none";
            }
        }
        function topFunction() {
            window.scrollTo({ top: 0, behavior: 'smooth' })
            document.documentElement.scrollTo({ top: 0, behavior: 'smooth' })
        }

        document.addEventListener("DOMContentLoaded", () => {
            document.querySelector('#mode').addEventListener('click',()=>{
                document.querySelector('html').classList.toggle('dark');
            })
        });


    </script>

    <nav class="navbar navbar-custom navbar-inverse navbar-fixed-top">
        <div class="container-fluid navbar-collapse collapse w-100">
            <a class="nav-item  active1"  href="{% rurl 'test-home' %}"><i class="fad fa-house-user"></i> Home</a>
            <a class="nav-item active1" style="padding-left: 20px" href="{% rurl 'test-upload' %}"><i class="fal fa-paper-plane"></i> Explore</a>
              <a class="nav-item  btn22 active22  active1" style="padding-left: 20px" href="{% rurl 'test-about' %}"><i class="far fa-info-square"></i> Documentation</a>
            <a class="nav-item active2" style="padding-left: 20px" href="#"><i class="fab fa-github-square"></i> Version 1.0.0</a>
        </div>
    </nav>

    <div id="wrapper" style="margin-top: 50px">

        <div id="mode" >
            <div class="dark">
                <svg aria-hidden="true" viewBox="0 0 512 512">
                    <title>lightmode</title>
                    <path fill="currentColor" d="M256 160c-52.9 0-96 43.1-96 96s43.1 96 96 96 96-43.1 96-96-43.1-96-96-96zm246.4 80.5l-94.7-47.3 33.5-100.4c4.5-13.6-8.4-26.5-21.9-21.9l-100.4 33.5-47.4-94.8c-6.4-12.8-24.6-12.8-31 0l-47.3 94.7L92.7 70.8c-13.6-4.5-26.5 8.4-21.9 21.9l33.5 100.4-94.7 47.4c-12.8 6.4-12.8 24.6 0 31l94.7 47.3-33.5 100.5c-4.5 13.6 8.4 26.5 21.9 21.9l100.4-33.5 47.3 94.7c6.4 12.8 24.6 12.8 31 0l47.3-94.7 100.4 33.5c13.6 4.5 26.5-8.4 21.9-21.9l-33.5-100.4 94.7-47.3c13-6.5 13-24.7.2-31.1zm-155.9 106c-49.9 49.9-131.1 49.9-181 0-49.9-49.9-49.9-131.1 0-181 49.9-49.9 131.1-49.9 181 0 49.9 49.9 49.9 131.1 0 181z"></path>
                </svg>
            </div>
            <div class="light">
                <svg aria-hidden="true" viewBox="0 0 512 512">
                    <title>darkmode</title>
                    <path fill="currentColor" d="M283.211 512c78.962 0 151.079-35.925 198.857-94.792 7.068-8.708-.639-21.43-11.562-19.35-124.203 23.654-238.262-71.576-238.262-196.954 0-72.222 38.662-138.635 101.498-174.394 9.686-5.512 7.25-20.197-3.756-22.23A258.156 258.156 0 0 0 283.211 0c-141.309 0-256 114.511-256 256 0 141.309 114.511 256 256 256z"></path>
                </svg>
            </div>
        </div>

        <div class="container">

            <section id="top" class="section docs-heading">

                <div class="row">
                    <div class="col-md-12">
                        <div class="big-title text-center">
                            <h1>DiNiro Documentation & Help</h1>
{#                            <p class="lead" style="text-align: right">DiNiro version 1.0.0</p>#}
                        </div>
                        <!-- end title -->
                    </div>
                    <!-- end 12 -->
                </div>
                <!-- end row -->

                <hr>

            </section>
            <!-- end section -->

            <div class="row">

                <div class="col-md-3">
                    <nav class="docs-sidebar" data-spy="affix" data-offset-top="300" data-offset-bottom="200" role="navigation">
                        <ul class="nav">
                            <li><a href="#line1">Home Page</a></li>
                            <li><a href="#line2">Explore Page</a></li>
                            <li><a href="#line3">Results Page</a></li>
                            <li><a href="#line8">Support Desk</a></li>
                            <li><a href="#line11">Copyright and license</a></li>
                        </ul>
                    </nav >
                </div>
                <div class="col-md-9">
                    <section class="welcome">

                        <div class="row">
                            <div class="col-md-12 left-align">
                               <h2 class="dark-text">Introduction<hr></h2>
                                <div class="row">

                                    <div class="col-md-12 full">
                                        <div class="intro1">
                                            <ul>
                                                <li><strong>Item Name : </strong>DiNiro Documentation</li>
                                                <li><strong>Item Version : </strong> v 1.0.0</li>
                                                <li><strong>Author  : </strong> <a href="https://www.zbh.uni-hamburg.de/en/personen/csb/moubounyt.html" target="_blank">Mhaned Oubounyt</a></li>
                                                <li><strong>Contact : </strong> <a href="mhaned.oubounyt@uni-hamburg.de" target="_blank">maned.oubounyt@uni-hamburg.de</a></li>
                                                <li><strong>License : </strong> <a href="#" target="_blank">Un License</a></li>
                                            </ul>
                                        </div>

                                        <hr>
                                        <div>
                                            <p style="color: black"> <strong>DiNiro</strong> (Differential co-expression regulatory networks analysis for Single-cell) is an interactive pipeline capable of extracting the co-expression regulatory
                                                networks and performing differential co-expression analysis from single-cell data. <strong>DiNiro</strong> is available as user-friendly open-source software and can be used interactively as a web-application at <strong><a href="https://exbio.wzw.tum.de/diniro" target="_blank">https://exbio.wzw.tum.de/diniro/</a></strong>.
                                            </p>
                                            <div class="intro2 clearfix">
                                                <p><i class="fa fa-exclamation-triangle"></i> <strong>Security Information</strong>
                                                    <br>All files are securely uploaded and stored on our servers.
                                                    <br>The data generated along the DiNiro pipeline are stored in folders with Unique User IDs that are bound to the session - so, if you close your browser, you can go back to your analysis using your unique URL.
                                                    <br>Whenever you start a new analysis (upload a file or try a test set), a unique session is created.
                                                    <br>All uploaded and DiNiro-generated data relating to analysis are deleted after 12 hours.
                                                </p>
                                            </div>

                                        </div>
                                    </div>

                                </div>
                                <!-- end row -->
                            </div>
                        </div>
                        <hr>
                    </section>

                    <section id="line1" class="section">

                        <div class="row">
                            <div class="col-md-12 left-align">
                                <h2 class="dark-text">Home Page<hr></h2>
                            </div>
                            <!-- end col -->
                        </div>
                        <!-- end row -->

                        <div class="row">
                            <div class="col-md-12">
                                <p style="color: black"> This is the landing page of the weeb-tool. They are three main buttons.
                                    Use [EXPLORE] to start the DiNiro interface and [DOCUMENTATION] to get to this page and tutorial to [VISUALIZE] the output of one of the
                                    precomputed examples.
                                </p>
                            </div>
                            <!-- end col -->
                        </div>
                        <!-- end row -->
                    <hr>
                    </section>
                    <!-- end section -->

                    <section id="line2" class="section">

                        <div class="row">
                            <div class="col-md-12 left-align">
                                <h2 class="dark-text">Explore Page<hr></h2>
                            </div>
                            <!-- end col -->
                        </div>
                        <!-- end row -->




                        <h4>Left view :</h4>

                        <div class="intro2 clearfix">
                            <p>[NEXT] button will take you to the next step if all the selections are made correctly. If not, DiNiro will throw a warning message.
                                                            <br> [BACK] button can be used to go back to the previous step but it will reset any input from the the current step.
                            </p>
                        </div>
                        <h6><strong>1. Input Data:</strong></h6>
                        <p> The input data to DiNiro must be a Scanpy AnnData object in H5AD file format. The H5AD file contains your analyzes, clusterings, and plots created with Scanpy. When using DiNiro, you will have access
                            to all the stored observations and variables in the uploaded AnnData object.
                            <br> To use DiNiro you have two options:
                            <br> 1) On the [Upload Files] card upload your desired H5AD file. You can additionally upload your Transcription Factors (TFs) of interest as a line by line text file (DiNiro will only use this TFs during analysis).
                            <br> Once you have selected the files to upload click on [UPLOAD] to save files and generate the single-cell data plot.
                            <br> 2) Alternatively, you can try out one of our four example data sets.  Click on the button [GO] Below the data set you would like to try.
                        </p>

                        <h6><strong>2. Plotting Choices:</strong></h6>
                        <p> The drop-down menu options are based on your pre-analysis and they will appear as you name them during scRNA-seq data processing with Scanpy.
                            <br> The single-cell plot is based on two main options: plotting-map and color-option:
                            <br> 1) Plotting-map is the dimensionality reduction method used for data visualization. You can use the [Type of plot] drop-down menu to choose among the already computed and stored plotting -maps in the Scanpy AnnData object. Typical methods are PCA, t-SNE, and UMAP.
                            <br> 2) Color-option is the coloring criteria to use for cell coloring such as cell type, tissue type, etc.
                            <br> [UPDATE PLOT] button allows you to update the plot and try a different combination of plotting -map and Color-option
                        </p>

                        <h6><strong>3. Selection:</strong></h6>
                        <p> DiNiro performs differential co-expression analysis between two samples. Thus, selecting samples is a crucial step..
                            <br> There are two types of selection methods to form your samples: CLUSTERS SELECTION and LASSO SELECTION:
                            <br> 1) CLUSTERS SELECTION is the straightforward method that allows you to form the samples from the single-cell clusters (a.k.a., cell types) available within your data. At least one cluster should be selected in order to form a sample, but multiple clusters can be also selected to form a sample. A cluster can be present in both samples.
                                    Use the [Select clusters to ADD to sample] drop-down menu to select the cluster and the [ADD] button to add it to the sample. Finally, use the [Type in an ID for sample] field to give your sample an ID (this will be useful later on when browsing the results).
                                    The same applies to making the second sample.
                                    When you click [CLEAR SELECTION] all the selections will be erased.

                            <br> 2) LASSO SELECTION comes in handy in case there are no clusters in the data in this case you can use this selection to manually add cells to each sample by drawing a circle around them.
                                    The first lasso selection will be automatically added to sample one and the selected cells will turn green, the text field will indicate the number of cells selected. The second selection will be added to sample two and the selected cell will turn red.
                                    Use [Type in an ID for sample] field to give your sample an ID (this will be useful later on when browsing the results).
                            <br> [RESET SELECTION] allows you to erase the selection and start over.
                        </p>

                        <h6><strong>4. Parameters:</strong></h6>
                        <p> For simplicity reasons, the parameter section is divided into two types of parameters: normal parameters and advanced parameters.
                        <div class="intro2 clearfix">
                            <p>
                                Warning: we advise the user to specify the advance parameter for more control over their analysis as this will eventually affect results and run time.
                            </p>
                        </div>

                            <br> The single-cell plot is based on two main options: plotting-map and color-option:
                            <br> 1) Parameters:
                                    The significance cutoff: We use a statistical test based on Copulas to compute the p-value for all the interactions found during the analysis (network edges) this cutoff is then used to keep the only significant ones. The default value is 0.05.
                                    Species selection: in case you have not uploaded your TFs of interest, DiNiro uses all available TFs for the species. User can select between [Mouse and Human].

                            <br> 2) Advanced Parameters (Hover over the parameter to see more details)



                        </p>

                        <h6><strong>5. Computing:</strong></h6>
                        <p> This view contains the computation progress bar and the URL  that you can use to retrieve your result later on.


                        </p>

                        <h4>Mideel view:</h4>

                        <h6><strong>1. Single-Cell Data Plot:</strong></h6>
                        <p>
                            This view is dedicated to interactively viewing the data. The user has control over the plotting map and the coloring criteria using the Plotting Choices view.
                        </p>

                        <h4>Right view:</h4>
                        <p>This view contains a summary of the data such as the number of cells, number of genes, and plotting maps.</p>



                    </section>
                    <!-- end section -->

                    <section id="line3" class="section" style="border-style: solid; border-width: 0.5mm; border-color: rgb(255, 255, 255);">

                        <div class="row">
                            <div class="col-md-12 left-align">
                                <h2 class="dark-text">Results Pages<hr></h2>
                            </div>
                            <!-- end col -->
                        </div>
                        <!-- end row -->

                        <p>
                            DiNro returns differentially co-expressed modules in the two input samples. A module is always formed from a TF and the differentially expressed genes regulated by this TF in both samples.
                            <br> DiNiro offers three different views of the results allowing the user more flexibility when interacting with the results.
                        </p>


                        <h4>TABLE :</h4>

                        <p>
                            You can view the results as a table of five columns as follow:
                        <br><strong>Rank :</strong> We use a differential expression score to rank the outputted modules the results can be also limited based on this rank (see parameters).
                        <br><strong>Module TF:</strong> The TF that regulates all the genes within the module.
                        <br><strong>Genes :</strong> The module gene.
                        <br><strong>GO Terms:</strong> The top significant go terms are based on the enrichment analysis of the module genes.
                        <br><strong>Module Network:</strong> Module networks are displayed in this column. TFs are the blue triangles and genes are the orange circle the direct edges are code colors based on in which sample the interaction was found. The network provides better visualization of the network.

                        </p>

                        <h4>NETWORK :</h4>

                        <p>
                         <br>The network allows you to visualize all modules together and spot the genes that are regulated by more than one TF and also search for some specific genes of interest (useful in the case of large networks).
                        <br>In top to left, you can type in or past the gene to search as comma-separated-values then click on [SEARCH] to start the search. If the genes are present in the network they will be colored in yellow.
                        <br>The download allows you to take a screenshot of the current view of the network. The network legends are displayed in the top left corner.
                        <br>In both the left and right bottom corner there are a bunch of buttons to interact with the network such as zoom in/out, etc.

                        </p>

                        <h4>REPORT :</h4>
                        <p>
                         <br> This view contains a dynamic Venn plot illustrating the distribution of the differentially expressed genes across the samples and also in form of a downloadable table.
                        <br> Similar to the network view here you can interact with the result in different ways. You can get information on the distribution of the differentially expressed genes across the samples. You can also search for genes of interest in the Venn plot using the search bar on top.


                        </p>
                        

                    </section>
                    <!-- end section -->


                    <section id="line8" class="section">

                        <div class="row">
                            <div class="col-md-12 left-align">
                                <h2 class="dark-text">Support Desk <hr></h2>
                            </div>
                            <!-- end col -->
                        </div>
                        <!-- end row -->

                        <div class="row">
                            <div class="col-md-12">

                                <strong>Contact:</strong>
                                <ul>
                                    <p>* If you want to contact us regarding DiNiro, please contact us in the following order:</p>
                                    <li>* Mhaned Oubounyt
                                        <br> oubounyt@wzw.tum.de</li>
                                    <li>* Jan Baumbach
                                          <br>  jan.baumbach@wzw.tum.de</li>
                                </ul>
                            </div>

                            <div class="col-md-6">
      
                            </div>
                        </div>
                        <!-- end row -->

                    </section>
                    <!-- end section -->

                    <section id="line11" class="section">

                        <div class="row">
                            <div class="col-md-12 left-align">
                                <h2 class="dark-text">Copyright and license <hr></h2>
                            </div>
                            <!-- end col -->
                        </div>
                        <!-- end row -->

                        <div class="row">
                            <div class="col-md-12">


                                <p></a>.</p>

                            </div>
                        </div>
                        <!-- end row -->

                    </section>
                    <!-- end section -->
                </div>
                <!-- // end .col -->

            </div>
            <!-- // end .row -->

        </div>
        <!-- // end container -->

    </div>
    <!-- end wrapper -->
    <!-- Footer -->
    <footer class="footer-custom bg-dark footer navbar-fixed-bottom">

      <!-- Copyright -->
{#      <div class="footer-copyright text-center py-3">© 2020 Copyright:#}
{#        <a href="https://mdbootstrap.com/"> MDBootstrap.com</a>#}
{#      </div>#}
            <div class="container">
                <p class="m-0 text-center text-white">Di-Niro | version 1.0.0 | &copy2021 <a style="text-decoration:none;" href="https://mdbootstrap.com/"> cosy.bio</a> </p>
            </div>

      <!-- Copyright -->

    </footer>
    <!-- Footer -->

    <script src="js/jquery.min.js"></script>
    <script src="js/bootstrap.min.js"></script>
    <script src="js/retina.js"></script>
    <script src="js/jquery.fitvids.js"></script>
    <script src="js/wow.js"></script>
    <script src="js/jquery.prettyPhoto.js"></script>

    <!-- CUSTOM PLUGINS -->
    <script src="js/custom.js"></script>
    <script src="js/main.js"></script>

    <script src="js/syntax-highlighter/scripts/shCore.js"></script>
    <script src="js/syntax-highlighter/scripts/shBrushXml.js"></script>
    <script src="js/syntax-highlighter/scripts/shBrushCss.js"></script>
    <script src="js/syntax-highlighter/scripts/shBrushJScript.js"></script>

</body>

</html>
